File: leetcode_ly/c++/DynamicProgramming/CountNumbersWithUniqueDigits.cpp
class Solution {
public:
    int countNumbersWithUniqueDigits(int n) {
        if(n == 0) return 1;
        int cnt = 0;
        for(int i = 1; i <= n; i++){
            cnt += get(i);
        }
        return cnt;
    }
    
    int get(int n){
        if(n == 1) return 10;
        int res = 9;
        for(int i = 2; i <= n; i++){
            res *= (11 - i);
        }
        return res;
    }
};


class Solution {
public:
    int countNumbersWithUniqueDigits(int n) {
        if(n == 0) return 1;
        if(n == 1) return 10;
        int cnt = 9;
        int res = 10;
        for(int i = 2; i <= n; i++){
            cnt *= (11 - i);
            res += cnt;
        }
        return res;
    }
};

File: leetcode_ly/java/Backtracking && DP/CountNumbersWithUniqueDigits.java
public class Solution {
    public int countNumbersWithUniqueDigits(int n) {
        if(n == 0) return 1;
        int res = 10;
        int digit  = 9;
        int remain = 9;
        while(n > 1 && remain > 0){
            digit *= remain;
            res += digit;
            n--;
            remain--;
        }
        return res;
    }
}œ